normalize term for instrumentations

toumorokoshi opened the pull request:
Fixes #476

A single term for instrumentation libraries would be beneficial, as
it reduces ambiguation. Removing synonyms "integration" and "adapter".

Integration doesn't imply some form of instrumentation, while adapter
could easily overlap with middleware or other concepts in existing
frameworks that will be instrumented.

arminru commented on specification/glossary.md:
```diff
@@ -23,16 +23,14 @@ Example: `org.mongodb.client`.
 <a name="instrumenting_library"></a>
 
-## Instrumenting Library
+## Instrumentation Library
```
Please add `Instrumenting Library` as a synonym then, pretty sure that's already being used at some places.

It would also make sense to link to the definition in `overview.md` here (or move it to the glossary entirely).

toumorokoshi replied:
Do you have a preference on moving the text in the overview? From what I see, the glossary itself is fairly terse, so I was wondering if it's appropriate to move the significant context that includes the recommendation of the naming convention for libraries.

arminru replied:
I think keeping a brief summary in the glossary and linking to the proper definitions and explanations given in other places is just fine.

arminru commented on specification/overview.md:
```diff
@@ -266,23 +266,24 @@ service).
-TODO: fill out as a result of SIG discussion.
+## Instrumentation Libraries
+
+A goal of the project is to enable opentelemetry instrumentation on every library
```
```suggestion
A goal of the project is to enable OpenTelemetry instrumentation on every library
```

The previous wording was different in terms of rationale however, there OTel was "just" a means to "make every library and application manageable out of the box".

arminru commented on specification/overview.md:
```diff
@@ -266,23 +266,24 @@ service).
+instrumentation libraries on these shared libraries and applications. These
+instrumentations can be wrapping library APIs that subscribe to library-specific
+callbacks or translators that convert telemetry exposed in other formats into
+opentelemetry models.
```
```suggestion
instrumentations wrap library APIs, subscribe to library-specific
callbacks or translate telemetry exposed in other formats into
OpenTelemetry models.
```

arminru commented on specification/overview.md:
```diff
@@ -266,23 +266,24 @@ service).
+callbacks or translators that convert telemetry exposed in other formats into
+opentelemetry models.
+
+When naming these instrumentations, it is recommended to follow the library and
+language standards. For instance, if an instrumentation is implemented as a
```
Are those supposed to be general standards or ones laid out by OpenTelemetry?

toumorokoshi replied:
I rephrased the original text, but I read the intention as the standards to pertain to the library that is being instrumented. e.g. if the common term is 'middleware', the instrumentation should follow that name, to be easily discoverable by the community of that library.

jkwatson commented on milestones.md:
```diff
@@ -137,9 +137,9 @@ and languages SDKs will implement OpenTelemetry collector exporter.
+By the end of year the basic language SDKs will be complete and we will begin
 stabilization work. Also OpenCensus can be switched to the OpenTelemetry SDK. As
-well as instrumentation adapters can be implemented. So we will have early
+well as instrumentations can be implemented. So we will have early
```
I think that `instrumentation` is the plural of `instrumentation` (no `s` required)

jkwatson commented on specification/error-handling.md:
```diff
@@ -44,7 +44,7 @@ And make a reasonable trade off of the SDK performance and fullness of type chec
 ## Self-diagnostics
 
-All OpenTelemetry libraries -- the API, SDK, exporters, instrumentation adapters, etc. -- are encouraged to expose self-troubleshooting metrics, spans, and other telemetry that can be easily enabled and filtered out by default.
+All OpenTelemetry libraries -- the API, SDK, exporters, instrumentations, etc. -- are encouraged to expose self-troubleshooting metrics, spans, and other telemetry that can be easily enabled and filtered out by default.
```
```suggestion
All OpenTelemetry libraries -- the API, SDK, exporters, instrumentation, etc. -- are encouraged to expose self-troubleshooting metrics, spans, and other telemetry that can be easily enabled and filtered out by default.
```

arminru commented on specification/overview.md:
```diff
@@ -266,23 +266,24 @@ service).
+"log appender" - it will probably be called an `appender`.
+
+If there is no established name - the recommendation is to call prefix packages
+with "instrumentation", followed by the instrumented library name itself.
```
```suggestion
If there is no established name, the recommendation is to prefix packages
with "opentelemetry-instrumentation", followed by the instrumented library name itself.
```

arminru commented on specification/overview.md:
```diff
@@ -266,23 +266,24 @@ service).
+If there is no established name - the recommendation is to call prefix packages
+with "instrumentation", followed by the instrumented library name itself.
+
+Here are some example library names that illustrate the opentelemetry
```
```suggestion
Here are some example library names that illustrate the OpenTelemetry
```

jkwatson commented on specification/overview.md:
```diff
@@ -266,23 +266,24 @@ service).
+A goal of the project is to enable opentelemetry instrumentation on every library
+and application out of the box. To achieve that goal, there is a need to author
+instrumentation libraries on these shared libraries and applications. These
+instrumentations can be wrapping library APIs that subscribe to library-specific
```
```suggestion
instrumentation libraries on these shared libraries and applications. This
instrumentation can be wrapping library APIs that subscribe to library-specific
```

jkwatson commented on specification/overview.md:
```diff
@@ -266,23 +266,24 @@ service).
+callbacks or translators that convert telemetry exposed in other formats into
+opentelemetry models.
+
+When naming these instrumentations, it is recommended to follow the library and
```
```suggestion
When naming instrumentation, it is recommended to follow the library and
```

carlosalberto commented:
@open-telemetry/specs-approvers please review/approve this PR ;)

carlosalberto commented:
@toumorokoshi Please resolve the (already resolved) comments so new reviewers can get a more condensed view ;)

thisthat commented on specification/glossary.md:
```diff
@@ -25,15 +25,15 @@ Example: `org.mongodb.client`.
 
 Denotes the library that provides the instrumentation for a given [Instrumented Library](#instrumented_library).
-*Instrumented Library* and *Instrumenting Library* may be the same library
+*Instrumented Library* and *Insrumentation Library* may be the same library
```
```suggestion
*Instrumented Library* and *Instrumentation Library* may be the same library
```

toumorokoshi commented:
Mentioned in the commits, but just as an FYI: I couldn't find a clear source to determine how to make "instrumentation" plural, so I renamed all referenced to "instrumentation libraries" instead.

SergeyKanzhelev commented on specification/overview.md:
```diff
@@ -262,22 +262,29 @@ service).
-Adapter" or simply "Adapter".
+## Instrumentation Libraries
+
+See [Instrumentation Library](glossary.md#instrumenting_library)
```
```suggestion
See [Instrumentation Library](glossary.md#instrumentation_library)
```

SergeyKanzhelev commented on specification/overview.md:
```diff
@@ -262,22 +262,29 @@ service).
+
+A goal of the project is to enable OpenTelemetry instrumentation on every library
+and application out of the box. To achieve that goal, there is a need to author
+instrumentation libraries on these shared libraries and applications. An
```
authoring an instrumentation library is a workaround while we are trying to instrument all libraries natively. This sentence implies opposite

SergeyKanzhelev commented on specification/overview.md:
```diff
@@ -262,22 +262,29 @@ service).
+callbacks or translators that convert telemetry exposed in other formats into
+OpenTelemetry models.
+
+If the library that is being instrumented has a convention around the naming of
```
this is confusing... "The library that is being instrumented"  can mean one of two:

- library was compiled with the reference to OpenTelemetry API
- library was instrumented by the means of providing an instrumentation library.

The "right" sentence here would be "If the library that is being instrumented with the instrumentation library has a convention..." Which begs the question - is the word adaptor so bad?

toumorokoshi replied:
Will update to disambiguate. Let me think on the wording a bit.

Regarding the choice of "instrumentation", see:
https://github.com/open-telemetry/opentelemetry-specification/issues/476#issuecomment-603314389

Among the conversation as well as in the spec SIG discussion, instrumentation seemed to resonate the most. Adapter can still overlap with a framework-specific middleware or other concept.

The other impetus I see at this point is that the concept of using these libraries to automatically instrumenting non-natively instrumented libraries is called "auto-instrumentation". As such I feel like there's a semantic grouping there.

SergeyKanzhelev commented:
CC: @mtwo as this is partially a PM question on how to explain customers the concept of instrumentation library

mtwo commented on specification/overview.md:
```diff
@@ -262,22 +262,26 @@ service).
+See [Instrumentation Library](glossary.md#instrumentation_library)
 
 The inspiration of the project is to make every library and application
 manageable out of the box by instrumenting it with OpenTelemetry. However on the
```
```suggestion
observable out of the box by having them call the OpenTelemetry APIs directly. However on the
```

mtwo commented on specification/overview.md:
```diff
@@ -262,22 +262,26 @@ service).
 
 The inspiration of the project is to make every library and application
 manageable out of the box by instrumenting it with OpenTelemetry. However on the
 way to this goal there will be a need to enable instrumentation by plugging
```
```suggestion
way to this goal there will be a need to enable instrumentation by connecting
```

mtwo commented on specification/overview.md:
```diff
@@ -262,22 +262,26 @@ service).
-called an `appender`, not an instrumentation adapter. However if there is no
-established name - the recommendation is to call packages "Instrumentation
-Adapter" or simply "Adapter".
+instrumentation libraries into the library of choice. These instrumentation 
```
```suggestion
instrumentation libraries into libraries that users want to observe. These instrumentation 
```

toumorokoshi commented:
@SergeyKanzhelev I've addressed your concern around the goal of OpenTelemetry.

@mtwo would you mind giving the PR another review? I rewrote the first paragraph significantly. After reading your comments I realized that the first paragraph could articulate several details more clearly.

mtwo commented on specification/error-handling.md:
```diff
@@ -44,7 +44,7 @@ And make a reasonable trade off of the SDK performance and fullness of type chec
 ## Self-diagnostics
 
-All OpenTelemetry libraries -- the API, SDK, exporters, instrumentation adapters, etc. -- are encouraged to expose self-troubleshooting metrics, spans, and other telemetry that can be easily enabled and filtered out by default.
+All OpenTelemetry libraries -- the API, SDK, exporters, an instrumentation, etc. -- are encouraged to expose self-troubleshooting metrics, spans, and other telemetry that can be easily enabled and filtered out by default.
```
Might want to make "an instrumentation" plural to fit with the others (?)

```suggestion
All OpenTelemetry libraries -- the API, SDK, exporters, instrumentations, etc. -- are encouraged to expose self-troubleshooting metrics, spans, and other telemetry that can be easily enabled and filtered out by default.
```

arminru commented on specification/glossary.md:
```diff
@@ -25,15 +25,15 @@ Example: `org.mongodb.client`.
 Example: `io.opentelemetry.contrib.mongodb`.
 
-Synonyms: *Instrumentation Library*, *Integration*.
+Synonyms: *Instrumentation Library*
```
```suggestion
Synonyms: *Instrumenting Library*
```

arminru commented on specification/glossary.md:
```diff
@@ -25,15 +25,15 @@ Example: `org.mongodb.client`.
-*Instrumented Library* and *Instrumenting Library* may be the same library
+*Instrumented Library* and *Instrumentation Library* may be the same library
 if it has built-in OpenTelemetry instrumentation.
 
```
```suggestion

See [Overview](overview.md#instrumentation_library) for a more detailed definition and naming guidelines.
```

arminru commented on specification/trace/api.md:
```diff
@@ -88,8 +88,7 @@ Some applications may use multiple `TracerProvider` instances, e.g. to provide
 
-- `name` (required): This name must identify the instrumentation library (also
-  referred to as integration, e.g. `io.opentelemetry.contrib.mongodb`) and *not*
+- `name` (required): This name must identify the instrumentation library (e.g. `io.opentelemetry.contrib.mongodb`) and *not*
```
```suggestion
- `name` (required): This name must identify the [instrumentation library](../overview.md#instrumentation_library) (e.g. `io.opentelemetry.contrib.mongodb`) and *not*
```

SergeyKanzhelev commented:
this seems to be mostly an alignment. not a new work. And we have enough sign offs. Merging

toumorokoshi commented:
Going to add some notes as a public system of record is good.

As of this post, the following languages have the following conventions:

# instrumentation
* dotnet
* java

# not instrumentation
* go (plugin?)
* js (plugin)
* ruby (adapters)
* python (extension)

dyladan commented:
JS maintainer here. Should we be renaming existing libraries to use the term "instrumentation", or just using that term in new libraries and keeping existing naming for packages already published? 

/cc @open-telemetry/javascript-maintainers @open-telemetry/go-maintainers @open-telemetry/ruby-maintainers @open-telemetry/python-maintainers @open-telemetry/technical-committee

MrAlias commented:
> JS maintainer here. Should we be renaming existing libraries to use the term "instrumentation", or just using that term in new libraries and keeping existing naming for packages already published?
> 
> /cc @open-telemetry/javascript-maintainers @open-telemetry/go-maintainers @open-telemetry/ruby-maintainers @open-telemetry/python-maintainers @open-telemetry/technical-committee

If there is going to be a push to do this, the time is best before the GA release. I'm interested what the @open-telemetry/technical-committee 's guidance here is.

toumorokoshi commented:
> JS maintainer here. Should we be renaming existing libraries to use the term "instrumentation", or just using that term in new libraries and keeping existing naming for packages already published?
> 
> /cc @open-telemetry/javascript-maintainers @open-telemetry/go-maintainers @open-telemetry/ruby-maintainers @open-telemetry/python-maintainers @open-telemetry/technical-committee

As @MrAlias mentioned, the best time to do a complete rename is before GA. I'll be filing tickets in all repos to start the discussion, but I think that's a SIG / TC decision. I'll see if I can bring it up in the next spec meeting as well.

As a data point, I'll be making a hard push to do a complete rename on the Python side, as well as adhering to the standard from now on.
